### PR TITLE
Tg 28 implement discriminator loss

### PR DESCRIPTION
**Pull request by Dmmc123**

### CODE REVIEW

File: src/models/modules/discriminator.py [hunk 1]
```diff
@@ -16,18 +17,26 @@ def __init__(self) -> None:
...
     def forward(
         self, visual_features: torch.Tensor, textual_features: torch.Tensor
     ) -> Any:
         """
         Fuse two types of features together to get output for feeding into the classification loss
         :param torch.Tensor visual_features:
-            Feature maps of an image after being processed by discriminator
-        :param torch.Tensor textual_features: Result of text encoder
-        :return: Logits for each word in the picture
+            Feature maps of an image after being processed by discriminator. Bx128x17x17
```
Comment (rohit901) [inline src/models/modules/discriminator.py:31]: after being processed by image encoder.

File: src/models/modules/discriminator.py [hunk 2]
```diff
@@ -41,33 +50,109 @@ def forward(
...
+        Initialize modules that reduce the features down to a set of logits
+
+        :param int n_words: Max length of a caption
+        """
+        super().__init__()
+        self.squisher = nn.Sequential(
+            conv2d(128, n_words),  # for reducing the channel dimension
+            nn.Conv2d(n_words, n_words, kernel_size=17),  # to reduce feature maps
+        )
+        # flattening BxLx1x1 into BxL
```
Comment (rohit901) [inline src/models/modules/discriminator.py:67]: might enhance clarity if you write this as 
(B, L, 1, 1) to (B * L)

File: src/models/modules/discriminator.py [hunk 3]
```diff
@@ -41,33 +50,109 @@ def forward(
...
+
+    def __init__(self, n_words: int) -> None:
+        super().__init__()
+        # recording value of L
+        self.n_words = n_words
+        # layer for increasing the number of textual channels
+        self.text_conv = conv1d(256, 17 * 17)
+        # for reduced textual + visual features down to 1x1 feature map
+        self.joint_conv = nn.Conv2d(128 + n_words, n_words, kernel_size=17)
+        # converting BxLx1x1 into BxL
```
Comment (rohit901) [inline src/models/modules/discriminator.py:96]: (B,L,1,1) ---> (B * L)
